design-proposals/orch-cli.md

```diff
@@ -0,0 +1,229 @@
+  describe the properties of the object to be created.
+
+- `delete` ... deletes an object from the Orchestrator.
+
+- `set` / `update` ... modifies one or more values in an existing object on the Orchestrator.
+
+- `apply` ... create or modify an object from a yaml or json specification, from stdin or from a local
+  file.
+
+##### Configuration and utility
+
+- `config` ... configures the CLI. For example, `cli config set endpoint https://my-orchestrator.com/`
+  TO-DO: This command inherited from the catalog CLI does not follow consistent syntax
+  and we should consider discarding it in favor treating configuration as any other verb-noun operation,
+  i.e. `cli update config --endpoint https://my-orchestrator.com/`.
+
+- `completion` ... return a script that can be used to configure autocompletion.
+  TO-DO: Consider treating completion as a verb/noun operation? i.e. `cli get completion bash`.
+
+- `upload <directory>` ... upload items to the Orchestrator. This is currently a catalog-specific
+  operation and uploads a directory of yaml files to the orchestrator.
+
+- `version` ... return the version number of the CLI.
+
+- `watch` ... watch an object in the Orchestrator using webhook or stream interface, returning changes
+  as they occur.
+  TO-DO: drop this?
+
+- `wipe` ... wipe all data in a project.
+  TO-DO: drop this? Should be implemented as part of `cli delete project`.
+
+#### Nouns
+
+This ADR does not seek to exhaustively list the set of nouns that are available on the CLI. Each
+component shall add their own nouns as appropriate. For example,
+
+- `cli list applications` ... The `application` noun applies to applications in the catalog.
+
+- `cli list registry` ... The `registry` noun applies to registries in the catalog.
+
+#### Subjects
+
+Subjects are additional information or context releated to a noun. Different nouns may support
+different subjects. For convenience sake, most nouns support a name subject and optionally for those
+objects that are versioned, a version subject. For example,
+
+- `cli get application nginx` ... returns the first nginx application, agnostic to the version.
+
+- `cli get application nginx 0.0.1` ... return exactly version 0.0.1 of the nginx application.
+
+If additional subjects beyond name and version are required, then they should be specified with addional
+options. For example,
+
+- `cli get application nginx --deployed true` ... get the nginx application that is currently deployed.
+
+#### Global Options
+
+- `-h` / `--help` ... return help. The help returned may be contextualized to the verb or noun that is
+  being used. For example `cli -h` returns global help whereas `cli create application -h` return help
+  that is relevant to creating applications.
+
+- `-n` / `--noauth` ... disables authentication checks, only useful in a devleopment environment.
+
+- `-p` / `--project` ... sets the project that will be used for the current verb. Required for verbs that
```

Comment:
shall we allow configuration of this ? meaning that somebody can set this once he has logged it. a step further will be automating this via the token/access. if the user has access to only one project no need to go through this step.

Reply:
I think we should also allow listing projects and their details and the org details.

Reply:
@Andrea-Campanella I think it may actually be part of the config file and you only need to use `-p` if the default is unset.
@SeanCondon I've avoided putting the list of nouns -- I feel like tenant-datamodel is a service like any other, and `list projects` and `create project` should be verb/noun combinations.